zndraw/server/tasks.py

```diff
@@ -186,20 +265,176 @@ def read_file(url: str, target: str):
+                if avail["available"]:
+                    print(f"Modifier is running on {pyclient}")
+                    msg = CeleryTaskData(
+                        target=pyclient,
+                        event="modifier:run",
+                        data={"params": data, "token": token},
+                    )
+                    vis.socket.emit("celery:task:results", asdict(msg))
+                    vis.socket.sleep(avail["timeout"])
+                    if not vis.available:
+                        print("modifier timed out")
+                        vis.socket.sleep(1)
+
+                        msg = CeleryTaskData(
+                            target=f"webclients_{vis.token}",
+                            event="message:alert",
+                            data=f"Modifier {NAME} did not finish in time.",
+                        )
+                        vis.socket.emit("celery:task:results", asdict(msg))
+
+                        msg = CeleryTaskData(
+                            target=f"webclients_{vis.token}",
+                            event="modifier:run:finished",
+                            data=None,
+                        )
+                        vis.socket.emit("celery:task:results", asdict(msg))
+
+                    vis.socket.sleep(10)
+                    vis.socket.disconnect()
+                    return
+
+            vis.socket.sleep(1)
+    if NAME in ROOM_MODIFIER_HOSTS.get(vis.token, []):
+        while True:
+            for pyclient in ROOM_MODIFIER_HOSTS[vis.token][NAME]:
+                msg = CeleryTaskData(target=pyclient, event="available", data=None)
+                avail = vis.socket.call("celery:task:call", asdict(msg))
+                print(f"Modifier {NAME} is available on {pyclient}: {avail}")
+                if avail["available"]:
+                    print(f"Modifier is running on {pyclient}")  #
+                    msg = CeleryTaskData(
+                        target=pyclient,
+                        event="modifier:run",
+                        data={"params": data, "token": token},
+                    )
+                    vis.socket.emit("celery:task:results", asdict(msg))
+                    vis.socket.sleep(avail["timeout"])
+                    if not vis.available:
+                        print("modifier timed out")
+                        vis.socket.sleep(1)
+
+                        msg = CeleryTaskData(
+                            target=f"webclients_{vis.token}",
+                            event="message:alert",
+                            data=f"Modifier {NAME} did not finish in time.",
+                        )
+                        vis.socket.emit("celery:task:results", asdict(msg))
+
+                        msg = CeleryTaskData(
+                            target=f"webclients_{vis.token}",
+                            event="modifier:run:finished",
+                            data=None,
+                        )
```

Comment:
If we're keeping this structure, we should try to reduce the code duplication. I'm struggling to see the differences in these outside getting the pyclient from different dictionaries.